File: scripts/import_issues_from_n8n.js
#!/usr/bin/env node
/**
 * Script to import issues from N8N agent JSON output into the Issues API
 * Usage: node import_issues_from_n8n.js <json_file> <api_token>
 */

const fs = require('fs');
const https = require('https');

const API_BASE_URL = 'https://naqashthaheem.com/api';

// Helper function to make HTTP requests
function makeRequest(method, url, headers, data = null) {
  return new Promise((resolve, reject) => {
    const urlObj = new URL(url);
    const options = {
      hostname: urlObj.hostname,
      port: urlObj.port || 443,
      path: urlObj.pathname + urlObj.search,
      method: method,
      headers: headers,
    };

    const req = https.request(options, (res) => {
      let body = '';
      res.on('data', (chunk) => {
        body += chunk;
      });
      res.on('end', () => {
        try {
          const parsed = JSON.parse(body);
          resolve({ status: res.statusCode, data: parsed });
        } catch (e) {
          resolve({ status: res.statusCode, data: body });
        }
      });
    });

    req.on('error', (error) => {
      reject(error);
    });

    if (data) {
      req.write(JSON.stringify(data));
    }
    req.end();
  });
}

// Fetch issue categories from API
async function fetchCategories(apiToken) {
  const headers = {
    'Content-Type': 'application/json',
    'Authorization': `Bearer ${apiToken}`,
  };

  const response = await makeRequest('GET', `${API_BASE_URL}/issue-categories`, headers);
  
  if (response.status !== 200) {
    console.error(`Error fetching categories: ${response.status}`);
    console.error(response.data);
    return {};
  }

  const categories = response.data.data || [];
  const categoryMap = {};

  categories.forEach((cat) => {
    categoryMap[cat.name] = cat.id;
  });

  console.log(`Fetched ${Object.keys(categoryMap).length} categories:`);
  Object.entries(categoryMap).forEach(([name, id]) => {
    console.log(`  - ${name}: ${id}`);
  });

  return categoryMap;
}

// Create an issue via API
async function createIssue(apiToken, issueData, categoryMap) {
  const headers = {
    'Content-Type': 'application/json',
    'Authorization': `Bearer ${apiToken}`,
  };

  // API now supports category_name directly, so we can use it as-is
  // Remove category_id if category_name is provided (API will handle the lookup)
  if (issueData.category_name) {
    delete issueData.category_id; // Remove category_id if category_name exists
  } else if (!issueData.category_id) {
    // If neither is provided, check if we can map from category_name
    const categoryName = issueData.category_name;
    if (categoryName && categoryMap[categoryName]) {
      issueData.category_id = categoryMap[categoryName];
      delete issueData.category_name;
    } else {
      issueData.category_id = null;
    }
  }

  // Ensure status is set
  issueData.status = 'resolved';

  const response = await makeRequest('POST', `${API_BASE_URL}/issues`, headers, issueData);

  if (response.status === 201) {
    return response.data.data;
  } else {
    console.error(`Error creating issue '${issueData.title}': ${response.status}`);
    console.error(response.data);
    return null;
  }
}

// Add resolution to an issue (admin only)
async function addResolution(apiToken, issueId, resolutionNotes) {
  const headers = {
    'Content-Type': 'application/json',
    'Authorization': `Bearer ${apiToken}`,
  };

  const response = await makeRequest(
    'POST',
    `${API_BASE_URL}/issues/${issueId}/status`,
    headers,
    {
      status: 'resolved',
      resolution_notes: resolutionNotes,
    }
  );

  if (response.status === 200) {
    return true;
  } else {
    console.error(`Error adding resolution to issue ${issueId}: ${response.status}`);
    console.error(response.data);
    return false;
  }
}

// Add a comment to an issue
async function addComment(apiToken, issueId, content) {
  const headers = {
    'Content-Type': 'application/json',
    'Authorization': `Bearer ${apiToken}`,
  };

  const response = await makeRequest(
    'POST',
    `${API_BASE_URL}/comments`,
    headers,
    {
      commentable_type: 'Issue',
      commentable_id: issueId,
      content: content,
    }
  );

  if (response.status === 201) {
    return true;
  } else {
    console.error(`Error adding comment to issue ${issueId}: ${response.status}`);
    console.error(response.data);
    return false;
  }
}

// Sleep helper
function sleep(ms) {
  return new Promise((resolve) => setTimeout(resolve, ms));
}

// Main import function
async function importIssues(jsonFile, apiToken) {
  // Load JSON file
  let data;
  try {
    const fileContent = fs.readFileSync(jsonFile, 'utf8');
    data = JSON.parse(fileContent);
  } catch (error) {
    console.error(`Error reading file '${jsonFile}':`, error.message);
    process.exit(1);
  }

  const issuesData = data.issues || [];
  if (!issuesData.length) {
    console.error("Error: No 'issues' array found in JSON");
    process.exit(1);
  }

  console.log(`Found ${issuesData.length} issues to import\n`);

  // Fetch categories
  console.log('Fetching categories...');
  const categoryMap = await fetchCategories(apiToken);
  if (!Object.keys(categoryMap).length) {
    console.error('Error: Could not fetch categories. Exiting.');
    process.exit(1);
  }

  console.log('\nStarting import...\n');

  // Import each issue
  let successCount = 0;
  let failedCount = 0;

  for (let idx = 0; idx < issuesData.length; idx++) {
    const item = issuesData[idx];
    const issueData = item.issue || {};
    const resolutionData = item.resolution || {};
    const commentsData = item.comments || [];

    if (!Object.keys(issueData).length) {
      console.log(`Issue ${idx + 1}: Skipping - no issue data`);
      failedCount++;
      continue;
    }

    const title = issueData.title || 'Untitled Issue';
    console.log(`Issue ${idx + 1}/${issuesData.length}: ${title}`);

    // Create issue
    const createdIssue = await createIssue(apiToken, { ...issueData }, categoryMap);

    if (!createdIssue) {
      console.log(`  ❌ Failed to create issue\n`);
      failedCount++;
      continue;
    }

    const issueId = createdIssue.id;
    console.log(`  ✅ Issue created (ID: ${issueId})`);

    // Add resolution
    const resolutionNotes = resolutionData.resolution_notes;
    if (resolutionNotes) {
      if (await addResolution(apiToken, issueId, resolutionNotes)) {
        console.log(`  ✅ Resolution added`);
      } else {
        console.log(`  ⚠️  Failed to add resolution`);
      }
    }

    // Add comments
    if (commentsData.length) {
      let commentCount = 0;
      for (const comment of commentsData) {
        const content = comment.content;
        if (content) {
          if (await addComment(apiToken, issueId, content)) {
            commentCount++;
          }
          await sleep(500); // Rate limiting
        }
      }

      if (commentCount > 0) {
        console.log(`  ✅ ${commentCount} comment(s) added`);
      }
    }

    successCount++;
    console.log();

    // Rate limiting
    await sleep(1000);
  }

  console.log('\n' + '='.repeat(50));
  console.log('Import complete!');
  console.log(`  ✅ Success: ${successCount}`);
  console.log(`  ❌ Failed: ${failedCount}`);
  console.log('='.repeat(50));
}

// Main execution
if (require.main === module) {
  const args = process.argv.slice(2);
  if (args.length !== 2) {
    console.log('Usage: node import_issues_from_n8n.js <json_file> <api_token>');
    console.log('\nExample:');
    console.log('  node import_issues_from_n8n.js issues.json your_api_token_here');
    process.exit(1);
  }

  const [jsonFile, apiToken] = args;
  importIssues(jsonFile, apiToken).catch((error) => {
    console.error('Fatal error:', error);
    process.exit(1);
  });
}

module.exports = { importIssues };

File: scripts/import_issues_from_n8n.py
#!/usr/bin/env python3
"""
Script to import issues from N8N agent JSON output into the Issues API
Usage: python import_issues_from_n8n.py <json_file> <api_token>
"""

import json
import sys
import requests
import time
from typing import Dict, List, Any

# API Configuration
API_BASE_URL = "https://naqashthaheem.com/api"
HEADERS = {
    "Content-Type": "application/json"
}

# Category name to ID mapping (you'll need to fetch this from API)
CATEGORY_MAP = {
    "Technical Support": None,  # Will be fetched from API
    "Programming Help": None,
    "System Administration": None,
    "Database Questions": None,
    "Networking & Security": None,
    "DevOps & CI/CD": None,
    "Frontend Development": None,
    "General IT Discussion": None,
    "Other": None,
}


def fetch_categories(api_token: str) -> Dict[str, int]:
    """Fetch issue categories from API and create name-to-ID mapping"""
    headers = {**HEADERS, "Authorization": f"Bearer {api_token}"}
    response = requests.get(f"{API_BASE_URL}/issue-categories", headers=headers)
    
    if response.status_code != 200:
        print(f"Error fetching categories: {response.status_code}")
        print(response.text)
        return {}
    
    categories = response.json().get("data", [])
    category_map = {}
    
    for cat in categories:
        category_map[cat["name"]] = cat["id"]
    
    print(f"Fetched {len(category_map)} categories:")
    for name, cat_id in category_map.items():
        print(f"  - {name}: {cat_id}")
    
    return category_map


def create_issue(api_token: str, issue_data: Dict[str, Any], category_map: Dict[str, int]) -> Dict[str, Any]:
    """Create an issue via API"""
    headers = {**HEADERS, "Authorization": f"Bearer {api_token}"}
    
    # API now supports category_name directly, so we can use it as-is
    # Remove category_id if category_name is provided (API will handle the lookup)
    if "category_name" in issue_data and issue_data["category_name"]:
        issue_data.pop("category_id", None)  # Remove category_id if category_name exists
    elif "category_id" not in issue_data:
        # If neither is provided, check if we can map from category_name
        category_name = issue_data.get("category_name")
        if category_name and category_name in category_map:
            issue_data["category_id"] = category_map[category_name]
            issue_data.pop("category_name", None)
        else:
            issue_data["category_id"] = None
    
    # Ensure status is set
    issue_data["status"] = "resolved"
    
    response = requests.post(
        f"{API_BASE_URL}/issues",
        headers=headers,
        json=issue_data
    )
    
    if response.status_code == 201:
        return response.json().get("data")
    else:
        print(f"Error creating issue '{issue_data.get('title')}': {response.status_code}")
        print(response.text)
        return None


def add_resolution(api_token: str, issue_id: int, resolution_notes: str) -> bool:
    """Add resolution to an issue (admin only)"""
    headers = {**HEADERS, "Authorization": f"Bearer {api_token}"}
    
    response = requests.post(
        f"{API_BASE_URL}/issues/{issue_id}/status",
        headers=headers,
        json={
            "status": "resolved",
            "resolution_notes": resolution_notes
        }
    )
    
    if response.status_code == 200:
        return True
    else:
        print(f"Error adding resolution to issue {issue_id}: {response.status_code}")
        print(response.text)
        return False


def add_comment(api_token: str, issue_id: int, content: str) -> bool:
    """Add a comment to an issue"""
    headers = {**HEADERS, "Authorization": f"Bearer {api_token}"}
    
    response = requests.post(
        f"{API_BASE_URL}/comments",
        headers=headers,
        json={
            "commentable_type": "Issue",
            "commentable_id": issue_id,
            "content": content
        }
    )
    
    if response.status_code == 201:
        return True
    else:
        print(f"Error adding comment to issue {issue_id}: {response.status_code}")
        print(response.text)
        return False


def import_issues(json_file: str, api_token: str):
    """Main function to import issues from JSON file"""
    # Load JSON file
    try:
        with open(json_file, 'r', encoding='utf-8') as f:
            data = json.load(f)
    except FileNotFoundError:
        print(f"Error: File '{json_file}' not found")
        sys.exit(1)
    except json.JSONDecodeError as e:
        print(f"Error: Invalid JSON in file '{json_file}': {e}")
        sys.exit(1)
    
    issues_data = data.get("issues", [])
    if not issues_data:
        print("Error: No 'issues' array found in JSON")
        sys.exit(1)
    
    print(f"Found {len(issues_data)} issues to import\n")
    
    # Fetch categories
    print("Fetching categories...")
    category_map = fetch_categories(api_token)
    if not category_map:
        print("Error: Could not fetch categories. Exiting.")
        sys.exit(1)
    
    print("\nStarting import...\n")
    
    # Import each issue
    success_count = 0
    failed_count = 0
    
    for idx, item in enumerate(issues_data, 1):
        issue_data = item.get("issue", {})
        resolution_data = item.get("resolution", {})
        comments_data = item.get("comments", [])
        
        if not issue_data:
            print(f"Issue {idx}: Skipping - no issue data")
            failed_count += 1
            continue
        
        title = issue_data.get("title", "Untitled Issue")
        print(f"Issue {idx}/{len(issues_data)}: {title}")
        
        # Create issue
        created_issue = create_issue(api_token, issue_data.copy(), category_map)
        
        if not created_issue:
            print(f"  ❌ Failed to create issue\n")
            failed_count += 1
            continue
        
        issue_id = created_issue.get("id")
        print(f"  ✅ Issue created (ID: {issue_id})")
        
        # Add resolution
        resolution_notes = resolution_data.get("resolution_notes")
        if resolution_notes:
            if add_resolution(api_token, issue_id, resolution_notes):
                print(f"  ✅ Resolution added")
            else:
                print(f"  ⚠️  Failed to add resolution")
        
        # Add comments
        if comments_data:
            comment_count = 0
            for comment in comments_data:
                content = comment.get("content")
                if content:
                    if add_comment(api_token, issue_id, content):
                        comment_count += 1
                    time.sleep(0.5)  # Rate limiting
        
            if comment_count > 0:
                print(f"  ✅ {comment_count} comment(s) added")
        
        success_count += 1
        print()
        
        # Rate limiting
        time.sleep(1)
    
    print(f"\n{'='*50}")
    print(f"Import complete!")
    print(f"  ✅ Success: {success_count}")
    print(f"  ❌ Failed: {failed_count}")
    print(f"{'='*50}")


if __name__ == "__main__":
    if len(sys.argv) != 3:
        print("Usage: python import_issues_from_n8n.py <json_file> <api_token>")
        print("\nExample:")
        print("  python import_issues_from_n8n.py issues.json your_api_token_here")
        sys.exit(1)
    
    json_file = sys.argv[1]
    api_token = sys.argv[2]
    
    import_issues(json_file, api_token)

